### DISCUSSION TITLE
Different highlight methods for focused and unfocused screen?

### DISCUSSION THREAD

**Discussion by hamensman** [qtile/qtile#2861] [Q&A]
In the `GroupBox` widget, is it possible to have different highlight methods for a screen which is **active and focused** vs a screen which is **active but unfocused**?

![Screenshot_2021-10-08_16-52-24](https://user-images.githubusercontent.com/76445071/136588171-bde8cdb6-0159-43f6-a1bb-f26e793bc80a.png)

For instance, have the focused screen highlighted by block (in this drawing, screen 2) and the unfocused screen highlighted by border (in this drawing, screen 1).

**Comment by elParaguayo** [comment]
Looks good.

No reason you couldn't do it.

How's your python?

**Reply by hamensman** [reply]
> How's your python?

Amateur 😂

I'm learning Python, so I chose Qtile as my window manager to expose myself to Python everyday.

But yeah - amateur...

**Reply by elParaguayo** [reply]
Nothing wrong in that, especially if you're willing to give things a try.

This would be a great way to learn a bit about python and qtile. Have a read of the GroupBox source code. You'd mainly just need to change the `draw` method, and maybe add some extra options to `defaults`.

I'm very happy to give you guidance if you want to have a go.

If not, I can add to my "to do" list but it's not a short list!

**Reply by hamensman** [reply]
Unfortunately, that's out of my depth at the moment 😥.

As I develop my proficiency though, I should be more competent in hacking on the source code.

**Reply by elParaguayo** [reply]
No problem at all and thanks for the suggestion.

**Reply by hamensman** [reply]
Thank **you**!

It might also be cool to be able to change the colours too:
![Screenshot_2021-10-08_16-52-24](https://user-images.githubusercontent.com/76445071/136605979-949bf218-1d17-4879-a9dc-8b2c476b7e5e.png)

Just noticed the `this_screen_border` key doesn't seem to do anything. If I read the documentation correctly, it should be essentially the unfocused counterpart to `this_current_screen_border`. If I, say, set `highlight_method` to block, I can clearly see the coloured block on the focused active screen, but no coloured block for an unfocused active screen. Maybe I'm understanding the purpose of it wrong...

**Reply by CSpencerND** [reply]
> Nothing wrong in that, especially if you're willing to give things a try.
> 
> This would be a great way to learn a bit about python and qtile. Have a read of the GroupBox source code. You'd mainly just need to change the `draw` method, and maybe add some extra options to `defaults`.
> 
> I'm very happy to give you guidance if you want to have a go.
> 
> If not, I can add to my "to do" list but it's not a short list!

I'd like to give this a shot and would appreciate some guidance if you could. I completed a python and course and want to get my hands dirty but trying to apply my knowledge feels challenging as it seems I would need to understand the inner-workings of qtile.
So far I added `other_screen_highlight_method` to the defaults. I'm thinking I should be able to add some kind of loop to the draw method to iterate over the screens with some logic like this ...
```
for screen in screens:
    if screen is active:
        draw this border
    else:
        draw other border
```
but I really don't know how to get the screens and check which one is focused or how to tell qtile which border to apply to which screen. Hopefully I'm on the right track and you can provide me some info that can help. Thanks.

**Reply by elParaguayo** [reply]
Nice!

Fortunately, you don't really need to understand the entire inner workings of Qtile, just how widgets work. Even more fortunately, there's a guide on this in the docs: http://docs.qtile.org/en/latest/manual/howto/widget.html

Most of the changes you'll need to make are in this section:
https://github.com/qtile/qtile/blob/7d2e9b1cc14b1f7ed7bf9e8627e004f725d998f0/libqtile/widget/groupbox.py#L363-L389

Line 363 means "if the group is currently shown on a screen"

Line 372 is "is the group the same as the group shown on the same screen as the widget"

Line 379 is "is the group on the currently focussed screen"

Using those, you should be able to check the screens you want without looping over them. What you're really trying to do is just insert a different highlight method into each of those scenarios identified above.

**Reply by CSpencerND** [reply]
Amazing, thank you so much! I've been watching this repo for a while, and this is the first thing I noticed that seemed feasible for me to tackle.

**Reply by hamensman** [reply]
> Using those, you should be able to check the screens you want without looping over them. What you're really trying to do is just insert a different highlight method into each of those scenarios identified above.

Do they also need to be defined in the `defaults`?

**Reply by elParaguayo** [reply]
You probably need something like a `other_screen_higlight_method` in defaults. That may be the only parameter you need to add.

**Reply by hamensman** [reply]
Here's what I've done with `custom_groupbox.py`:  [custom_groupbox.txt](https://github.com/qtile/qtile/files/7330023/custom_groupbox.txt)

I've replaced:
```
import itertools
from functools import partial
from typing import Any, List, Tuple

from libqtile import bar, hook
from libqtile.widget import base
```
with this: `from libqtile import widget`

Added `other_screen_highlight_method` to defaults from `class CustomGroupBox`:
```
    defaults = [
        (
            "other_screen_highlight_method",
            "border",
            "Method of highlighting unfocused screen('border', 'block', 'text', or 'line')"
            "Uses `*_border` color settings"
        ),
...
```

And edited one of the parts from the section you highlighted:
```
if g.screen:
                if self.highlight_method == 'text':
                    border = None
                    text_color = self.this_current_screen_border
                else:
                    if self.block_highlight_text_color:
                        text_color = self.block_highlight_text_color
                    if self.bar.screen.group.name == g.name:
                        if self.qtile.current_screen == self.bar.screen:
                            border = self.other_screen_highlight_method #THIS ONE!
                            to_highlight = True
                        else:
                            border = self.this_screen_border
                    else:
                        if self.qtile.current_screen == g.screen:
                            border = self.other_current_screen_border
                        else:
                            border = self.other_screen_border
            elif self.group_has_urgent(g) and \
                    self.urgent_alert_method in ('border', 'block', 'line'):
                border = self.urgent_border
                if self.urgent_alert_method == 'block':
                    is_block = True
                elif self.urgent_alert_method == 'line':
                    is_line = True
            else:
                border = None
```
___
Now, I'm getting the obvious error of `base` not being defined: [qtile.log](https://github.com/qtile/qtile/files/7330068/qtile.log). But I'm not sure what to replace all mentions of `base` with...

**Reply by elParaguayo** [reply]
You don't need to copy everything from the current widget, only those bits you want to change.

Like your `CustomWindowName` widget, you just need to:
1) subclass `widget.GroupBox` (rather than copying the classes subclassed by `GroupBox`)
2) set `defaults`
3) define an `__init__` method to add your defaults
4) copy the `draw` method and make your changes in that.

You shouldn't need `base` anywhere.

**Reply by CSpencerND** [reply]
I think I was on to something, but I'm out of town for the next week, and I forgot to take an HDMI cable so I can't test it 😖. Here's what I got for now. You'll probably solve this before I get back. 
```python
    def draw(self):
        self.drawer.clear(self.background or self.bar.background)

        offset = self.margin_x
        for i, g in enumerate(self.groups):
            to_highlight = False
            # resets the highlight method every time the loop restarts
            is_block = False
            is_line = False

            bw = self.box_width([g])

            if self.group_has_urgent(g) and self.urgent_alert_method == "text":
                text_color = self.urgent_text
            elif g.windows:
                text_color = self.active
            else:
                text_color = self.inactive

            if g.screen:
                if self.highlight_method == 'text':
                    border = None
                    text_color = self.this_current_screen_border
                else:
                    if self.block_highlight_text_color:
                        text_color = self.block_highlight_text_color
                    if self.bar.screen.group.name == g.name:
                        if self.qtile.current_screen == self.bar.screen:
                            border = self.this_current_screen_border
                            to_highlight = True
                            # uses the normal highlight method if group is on screen with widget 
                            is_block = (self.highlight_method == 'block')
                            is_line = (self.highlight_method == 'line')
                        else:
                            border = self.this_screen_border
                    else:
                        if self.qtile.current_screen == g.screen:
                            border = self.other_current_screen_border
                            to_highlight = True
                            # uses the new highlight method if group is on other screen
                            is_block = (self.other_screen_highlight_method == 'block')
                            is_line = (self.other_screen_highlight_method == 'line')
                        else:
                            border = self.other_screen_border
```

**Reply by CSpencerND** [reply]
It should probably have a conditional to check `if other_screen_highlight_method:` to make it optional to have different borders and whether or not `to_highlight`.

**Reply by hamensman** [reply]
I feel like I'm so close, but not quite there yet: [qtile.log](https://github.com/qtile/qtile/files/7331673/qtile.log) and [custom_groupbox.txt](https://github.com/qtile/qtile/files/7331687/custom_groupbox.txt)

> subclass `widget.GroupBox` (rather than copying the classes subclassed by `GroupBox`)

If I understand this correctly, this means `CustomGroupBox` inherits the main features of the **normal** `GroupBox` widget? Which makes sense now, since that's way more efficient.

**Reply by hamensman** [reply]
@CSpencerND What does your `__init__` look like (if you're able to provide at this time)?

**Reply by CSpencerND** [reply]
> @CSpencerND What does your `__init__` look like (if you're able to provide at this time)?

I actually haven't touched it.

**Reply by hamensman** [reply]
> I actually haven't touched it.

And do you have any qtile error logs? Would be quite interested what it has written for you.

**Reply by elParaguayo** [reply]
When you subclass something but want to keep a method and add things to it, you don't need to rewrite the whole thing. So, for `__init__` you'd do this
``` python
class CustomGroupBox(widget.GroupBox):
    """
    A widget that graphically displays the current group.
    All groups are displayed by their label.
    If the label of a group is the empty string that group will not be displayed.
    """
    defaults = [
        (
            "other_screen_highlight_method",
            "border",
            "Method of highlighting unfocused screen('border', 'block', 'text', or 'line')"
            "Uses `*_border` color settings"
        ),
    ]

    def __init__(self, **config):
        # This line runs the __init__ method in widget.GroupBox
        widget.GroupBox.__init__(self, **config)
        self.add_defaults(CustomGroupBox.defaults)
```

**Reply by hamensman** [reply]
I see. Thanks for clarifying that, @elParaguayo.

Now, almost the entire bar is blank:
[qtile.log](https://github.com/qtile/qtile/files/7332191/qtile.log)
![Screenshot_2021-10-12_18-52-41](https://user-images.githubusercontent.com/76445071/137005769-f47100b7-6eb1-4e57-bd3f-27ba5f2b79c7.png)

I definitely messed up the `draw` method.

**Reply by elParaguayo** [reply]
You should get used to reading logs. Python error messages are usually very helpful.

In your case, the error says:
```
AttributeError: CustomGroupBox has no attribute: other_highlight_method
```
Have a look at the name you set in `defaults`...

**Reply by hamensman** [reply]
> You should get used to reading logs. Python error messages are usually very helpful.

Absolutely.

I've sorted the attribute discrepancy.
The recurring error is now a `ValueError` of some sort:
```
vals = [int(i, 16) for i in (x[0:2], x[2:4], x[4:6])]
ValueError: invalid literal for int() with base 16: 'bo'
```
[qtile.log](https://github.com/qtile/qtile/files/7332334/qtile.log)

I have no clue what this is specifically pointing to though...

**Reply by CSpencerND** [reply]
> > I actually haven't touched it.
> 
> And do you have any qtile error logs? Would be quite interested what it has written for you.

It's gonna be another two hours before I can get ya that.

**Reply by elParaguayo** [reply]
This one is harder. You need to look at more of the error message so you can see where the error is originating. It's in one of the colour parsing functions so I'm guessing `self.border_color` has a value of `None`.

**Reply by hamensman** [reply]
`/usr/lib/python3.9/site-packages/libqtile/utils.py` appears to be one suspect, complaining about `has_transparency`, `has_alpha`, `rgb` and `<listcomp>`.

**Reply by elParaguayo** [reply]
Yes, because it's received a value it shouldn't have. It's the value of `self.border_color` that's the problem by the look of it.

**Reply by elParaguayo** [reply]
Hang on, I've re-read the error. You've set `border_color` to the value of `other_screen_highlight_method`. It's trying to split a hex colour but you've given it "border" and it can't convert that into rgb.

**Reply by hamensman** [reply]
Thanks for pointing that silly error on my end 😆

Nothing in the error logs, but it isn't producing the desired result. What's happening is the `other_screen_highlight_method` is taking over the `highlight_method`:

![Screenshot_2021-10-12_21-03-29](https://user-images.githubusercontent.com/76445071/137021560-732904ec-b5ea-4c7c-a890-500c645792a4.png)

From reading the `if g.screen` section, my guess is it's to do with `this_current_screen_border`. `CustomGroupBox` appears to essentially be following the same protocol as the normal `GroupBox` widget. It's affecting the focused screen only. Hence, overwriting the normal `highlight_method` variable.

**Reply by elParaguayo** [reply]
Share the code for your custom widget and I'll take a look (although I'm running with just 1 screen at the moment)

**Reply by hamensman** [reply]
[custom_groupbox.txt](https://github.com/qtile/qtile/files/7332898/custom_groupbox.txt)

**Reply by elParaguayo** [reply]
It looks like you've replaced current_screen bits with other_screen bits. That won't work. You've got to do both i.e. you keep the old code but add the extra bits for the scenario when the other_current_screen is being displayed and, at that point, check what highlight method is needed for that group.

**Reply by CSpencerND** [reply]
> > I actually haven't touched it.
> 
> And do you have any qtile error logs? Would be quite interested what it has written for you.

I have no errors in my qtile.log at all, but since I don't have another screen right now, it is not a controlled experiment, so it may not be indicative of anything.

**Reply by hamensman** [reply]
@CSpencerND I replaced my `draw` method with your's just to see what would happen.

I also don't see anything in `qtile.log`. I do, however, see the screen widgets completely blanked out:
![Screenshot_2021-10-12_21-47-08](https://user-images.githubusercontent.com/76445071/137027038-6336f94a-39da-4eea-a787-38cdfc965fe1.png)

**Reply by CSpencerND** [reply]
Yeesh. Well, nothing I can do at the moment. Hopefully I can get an HDMI cable within the next couple days. 
Just to be sure: Do you have both `highlight_method` and `other_screen_highlight_method` defined in the defaults?

**Reply by hamensman** [reply]
I did not.
I added `highlight_method` back in, but same result unfortunately: [custom_groupbox.txt](https://github.com/qtile/qtile/files/7333465/custom_groupbox.txt)

**Reply by hamensman** [reply]
I think I've been using the world "screen" (in this context) incorrectly.

If you look at my screenshots (showing 1, 2, 3, etc.), in my head I'd think of it as, say, "workspace 1" or "virtual desktop 1" (in the same physical display). So judging from the very name of the widget in question, as well as the terms used in the `GroupBox` code, this would be - in qtile terms - "group 1"?

**Comment by diego-rapoport** [comment]
Do you mean something like **that**?
![08-10-2021-13h53min25](https://user-images.githubusercontent.com/66575787/136594882-26e9a9b8-2d9b-4da3-8381-9b2cad27135d.jpg)

The first two are backgrounds for screen 1 and screen 2. Both active.
Screen 3 and 4 are being used, but not active, and the others are inactive.
I'm using all colors with pywal.

**Reply by elParaguayo** [reply]
Not quite, you've got the `block` highlight method for both screens with different colours. @hamensman, I think, wants to have `block` for the active screen and `border` for the inactive screen.

**Reply by hamensman** [reply]
@diego-rapoport Just so I'm making sense of how the code is written, are you using a dual(multi)-monitor setup?

**Reply by diego-rapoport** [reply]
Indeed I am. Two monitors specifically.

**Reply by hamensman** [reply]
I thought so.

Can you post basically the same screenshot, but with 3 and 4 inactive (just like 5, 6, 7, 8, 9)? I want to illustrate a point with that.

**Reply by diego-rapoport** [reply]
Like that?

![13-10-2021-10h13min34](https://user-images.githubusercontent.com/66575787/137139648-b9343bb1-4e97-4583-9040-89f75c199339.jpg)

**Reply by hamensman** [reply]
Perfect!

I tried (just to test things out) to produce that with the existing `GroupBox` parameters, **but** with just one monitor. So imagine this bar indicates just one monitor, and that while both 1 and 2 are active, you are just looking at 2 right now.
![137139648-b9343bb1-4e97-4583-9040-89f75c199339](https://user-images.githubusercontent.com/76445071/137139900-730fcbe9-6c4f-4f08-84e9-a5fc60ce928d.jpg)

I have not been able to do that. I thought setting `this_screen_border` would produce this, but it did not.

I was testing this because I thought I should be able to have `GroupBox` widget apply the same highlight method before even thinking about changing them.

Does that make sense?

**Reply by diego-rapoport** [reply]
I couldn't understand what you're trying to do, but that's an example of my setting with only one monitor:

![13-10-2021-10h46min27](https://user-images.githubusercontent.com/66575787/137147277-b3210258-2cfe-413a-ac61-3f8bc2b3bb19.jpg)

This way 1 and 2 are active, but I'm looking at the desktop 3.

**Reply by hamensman** [reply]
OK. Here, I essentially want 1 and 2 to have a highlight too.

**Comment by CSpencerND** [comment]
@hamensman I was able to temporarily get a cable to test things. It works as expected with one caveat.
```python
import itertools
from functools import partial
from typing import Any, List, Tuple

from libqtile import bar, hook
from libqtile.widget import base


class _GroupBase(base._TextBox, base.PaddingMixin, base.MarginMixin):
    defaults = [
        ("borderwidth", 3, "Current group border width"),
        ("center_aligned", True, "center-aligned group box"),
    ]  # type: List[Tuple[str, Any, str]]

    def __init__(self, **config):
        base._TextBox.__init__(self, width=bar.CALCULATED, **config)
        self.add_defaults(_GroupBase.defaults)
        self.add_defaults(base.PaddingMixin.defaults)
        self.add_defaults(base.MarginMixin.defaults)

    def box_width(self, groups):
        width, _ = self.drawer.max_layout_size(
            [i.label for i in groups],
            self.font,
            self.fontsize
        )
        return width + self.padding_x * 2 + self.borderwidth * 2

    def _configure(self, qtile, bar):
        base._Widget._configure(self, qtile, bar)

        if self.fontsize is None:
            calc = self.bar.height - self.margin_y * 2 - \
                self.borderwidth * 2 - self.padding_y * 2
            self.fontsize = max(calc, 1)

        self.layout = self.drawer.textlayout(
            "",
            "ffffff",
            self.font,
            self.fontsize,
            self.fontshadow
        )
        self.setup_hooks()

    def setup_hooks(self):
        def hook_response(*args, **kwargs):
            self.bar.draw()
        hook.subscribe.client_managed(hook_response)
        hook.subscribe.client_urgent_hint_changed(hook_response)
        hook.subscribe.client_killed(hook_response)
        hook.subscribe.setgroup(hook_response)
        hook.subscribe.group_window_add(hook_response)
        hook.subscribe.current_screen_change(hook_response)
        hook.subscribe.changegroup(hook_response)

    def drawbox(self, offset, text, bordercolor, textcolor, highlight_color=None,
                width=None, rounded=False, block=False, line=False, highlighted=False):
        self.layout.text = text
        self.layout.font_family = self.font
        self.layout.font_size = self.fontsize
        self.layout.colour = textcolor
        if width is not None:
            self.layout.width = width
        if line:
            pad_y = [
                (self.bar.height - self.layout.height - self.borderwidth) / 2,
                (self.bar.height - self.layout.height + self.borderwidth) / 2
            ]
        else:
            pad_y = self.padding_y

        if bordercolor is None:
            # border colour is set to None when we don't want to draw a border at all
            # Rather than dealing with alpha blending issues, we just set border width
            # to 0.
            border_width = 0
            framecolor = self.background or self.bar.background
        else:
            border_width = self.borderwidth
            framecolor = bordercolor

        framed = self.layout.framed(
            border_width,
            framecolor,
            0,
            pad_y,
            highlight_color
        )
        y = self.margin_y
        if self.center_aligned:
            for t in base.MarginMixin.defaults:
                if t[0] == 'margin':
                    y += (self.bar.height - framed.height) / 2 - t[1]
                    break
        if block and bordercolor is not None:
            framed.draw_fill(offset, y, rounded)
        elif line:
            framed.draw_line(offset, y, highlighted)
        else:
            framed.draw(offset, y, rounded)


class GroupBox(_GroupBase):
    """
    A widget that graphically displays the current group.
    All groups are displayed by their label.
    If the label of a group is the empty string that group will not be displayed.
    """
    orientations = base.ORIENTATION_HORIZONTAL
    defaults = [
        ("block_highlight_text_color", None, "Selected group font colour"),
        ("active", "FFFFFF", "Active group font colour"),
        ("inactive", "404040", "Inactive group font colour"),
        (
            "highlight_method",
            "border",
            "Method of highlighting ('border', 'block', 'text', or 'line')"
            "Uses `*_border` color settings"
        ),
        (
            "other_screen_highlight_method",
            "border",
            "Method of highlighting for non-focused screen \
            ('border', 'block', 'text', or 'line')"
            "Uses `*_border` color settings"
        ),
        ("rounded", True, "To round or not to round box borders"),
        (
            "this_current_screen_border",
            "215578",
            "Border or line colour for group on this screen when focused."
        ),
        (
            "this_screen_border",
            "215578",
            "Border or line colour for group on this screen when unfocused."
        ),
        (
            "other_current_screen_border",
            "404040",
            "Border or line colour for group on other screen when focused."
        ),
        (
            "other_screen_border",
            "404040",
            "Border or line colour for group on other screen when unfocused."
        ),
        (
            "highlight_color",
            ["000000", "282828"],
            "Active group highlight color when using 'line' highlight method."
        ),
        (
            "urgent_alert_method",
            "border",
            "Method for alerting you of WM urgent "
            "hints (one of 'border', 'text', 'block', or 'line')"
        ),
        ("urgent_text", "FF0000", "Urgent group font color"),
        ("urgent_border", "FF0000", "Urgent border or line color"),
        (
            "disable_drag",
            False,
            "Disable dragging and dropping of group names on widget"
        ),
        ("invert_mouse_wheel", False, "Whether to invert mouse wheel group movement"),
        ("use_mouse_wheel", True, "Whether to use mouse wheel events"),
        (
            "visible_groups",
            None,
            "Groups that will be visible. "
            "If set to None or [], all groups will be visible."
            "Visible groups are identified by name not by their displayed label."
        ),
        (
            "hide_unused",
            False,
            "Hide groups that have no windows and that are not displayed on any screen."
        ),
        (
            "spacing",
            None,
            "Spacing between groups"
            "(if set to None, will be equal to margin_x)")
    ]

    def __init__(self, **config):
        _GroupBase.__init__(self, **config)
        self.add_defaults(GroupBox.defaults)
        if self.spacing is None:
            self.spacing = self.margin_x
        self.clicked = None
        self.click = None

        default_callbacks = {'Button1': self.select_group}
        if self.use_mouse_wheel:
            default_callbacks.update({
                'Button5' if self.invert_mouse_wheel else 'Button4': self.prev_group,
                'Button4' if self.invert_mouse_wheel else 'Button5': self.next_group,
            })
        self.add_callbacks(default_callbacks)

    @property
    def groups(self):
        """
        returns list of visible groups.
        The existing groups are filtered by the visible_groups attribute and
        their label. Groups with an empty string as label are never contained.
        Groups that are not named in visible_groups are not returned.
        """
        if self.hide_unused:
            if self.visible_groups:
                return [g for g in self.qtile.groups
                        if g.label and (g.windows or g.screen) and
                        g.name in self.visible_groups]
            else:
                return [g for g in self.qtile.groups if g.label and
                        (g.windows or g.screen)]
        else:
            if self.visible_groups:
                return [g for g in self.qtile.groups
                        if g.label and g.name in self.visible_groups]
            else:
                return [g for g in self.qtile.groups if g.label]

    def get_clicked_group(self):
        group = None
        new_width = self.margin_x - self.spacing / 2.0
        width = 0
        for g in self.groups:
            new_width += self.box_width([g]) + self.spacing
            if width <= self.click <= new_width:
                group = g
                break
            width = new_width
        return group

    def button_press(self, x, y, button):
        self.click = x
        _GroupBase.button_press(self, x, y, button)

    def next_group(self):
        group = None
        current_group = self.qtile.current_group
        i = itertools.cycle(self.qtile.groups)
        while next(i) != current_group:
            pass
        while group is None or group not in self.groups:
            group = next(i)
        self.go_to_group(group)

    def prev_group(self):
        group = None
        current_group = self.qtile.current_group
        i = itertools.cycle(reversed(self.qtile.groups))
        while next(i) != current_group:
            pass
        while group is None or group not in self.groups:
            group = next(i)
        self.go_to_group(group)

    def select_group(self):
        self.clicked = None
        group = self.get_clicked_group()
        if not self.disable_drag:
            self.clicked = group
        self.go_to_group(group)

    def go_to_group(self, group):
        if group:
            if self.bar.screen.group != group or not self.disable_drag:
                self.bar.screen.set_group(group, warp=False)
            else:
                self.bar.screen.toggle_group(group, warp=False)

    def button_release(self, x, y, button):
        self.click = x
        if button not in (5, 4):
            group = self.get_clicked_group()
            if group and self.clicked:
                group.cmd_switch_groups(self.clicked.name)
                self.clicked = None

    def calculate_length(self):
        width = self.margin_x * 2 + (len(self.groups) - 1) * self.spacing
        for g in self.groups:
            width += self.box_width([g])
        return width

    def group_has_urgent(self, group):
        return len([w for w in group.windows if w.urgent]) > 0

    def draw(self):
        self.drawer.clear(self.background or self.bar.background)

        offset = self.margin_x
        for i, g in enumerate(self.groups):
            to_highlight = False
            is_block = False
            is_line = False

            bw = self.box_width([g])

            if self.group_has_urgent(g) and self.urgent_alert_method == "text":
                text_color = self.urgent_text
            elif g.windows:
                text_color = self.active
            else:
                text_color = self.inactive

            # if the group is currently shown on a screen
            if g.screen:
                if self.highlight_method == 'text':
                    border = None
                    text_color = self.this_current_screen_border
                else:
                    if self.block_highlight_text_color:
                        text_color = self.block_highlight_text_color

                    # is the group the one shown on the screen with the widget
                    if self.bar.screen.group.name == g.name:
                        if self.qtile.current_screen == self.bar.screen:
                            border = self.this_current_screen_border
                            to_highlight = True
                            is_block = (self.highlight_method == 'block')
                            is_line = (self.highlight_method == 'line')
                        else:
                            border = self.this_screen_border
                    else:

                        # is the group on the currently focused screen
                        if self.qtile.current_screen == g.screen:
                            border = self.other_current_screen_border
                            to_highlight = True
                            is_block = (self.other_screen_highlight_method == 'block')
                            is_line = (self.other_screen_highlight_method == 'line')
                        else:
                            border = self.other_screen_border

            elif self.group_has_urgent(g) and \
                    self.urgent_alert_method in ('border', 'block', 'line'):
                border = self.urgent_border
                if self.urgent_alert_method == 'block':
                    is_block = True
                elif self.urgent_alert_method == 'line':
                    is_line = True
            else:
                border = None

            self.drawbox(
                offset,
                g.label,
                border,
                text_color,
                highlight_color=self.highlight_color,
                width=bw,
                rounded=self.rounded,
                block=is_block,
                line=is_line,
                highlighted=to_highlight
            )
            offset += bw + self.spacing
        self.drawer.draw(offsetx=self.offset, width=self.width)
```

**Reply by hamensman** [reply]
You can take this, which inherits the features of the normal GroupBox widget without having to write all the other stuff, as @elParaguayo has kindly explained: [custom_groupbox.txt](https://github.com/qtile/qtile/files/7340843/custom_groupbox.txt).

**Comment by CSpencerND** [comment]
When switching screens with `cmd_next_screen()`, the highlighting does NOT work the same as when switching groups, so there is one more piece of the puzzle missing.

**Reply by hamensman** [reply]
You're absolute right.

It's certainly good that you can change the highlight method of focused window in another display though. For that, you've pretty much done that perfectly.

But since clearing up the terms, it seems highlighting for **groups** would be quite tricky. Although, it would essentially be the highlight version of the `active` parameter, insofar as `active` changes the text colour.

In that sense, we would have to create at least 2 additional parameters. Something like `active_highlight_method` and `active_highlight_color` (I suppose`active` stays as it is so as to not break anything)

**Reply by CSpencerND** [reply]
Disregard what I said about it not working with `cmd_next_screen()`. All you have to do to make this work is define `other_screen_highlight_method='line',` or `block` in `config.py`.

**Reply by CSpencerND** [reply]
Oh boy. I just realized `other_screen_highlight_method` is redundant for the way we're using it, but if I remove it, if forces users to have the default `border` highlight method for inactive screens. I'm going to work on optimizing the code for different use cases before submitting a pr.

**Reply by CSpencerND** [reply]
@elParaguayo I submitted a pr for this. Everything works as expected on my end.

**Reply by elParaguayo** [reply]
Thanks. Will take a look.

**Comment by elParaguayo** [comment]
Move to #4365
